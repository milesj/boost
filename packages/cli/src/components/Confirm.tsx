/* eslint-disable no-console */

import React, { useCallback, useEffect, useState } from 'react';
import { useFocus } from 'ink';
import { msg } from '../translate';
import { Prompt, PromptProps } from './internal/Prompt';
import { Selected } from './internal/Selected';
import { Style } from './Style';

export interface ConfirmProps extends Omit<PromptProps<boolean>, 'validate'> {
	/** Error message to display when an invalid character is pressed. Defaults to a custom message. */
	invalidError?: string;
	/** Character that triggers a falsy state when pressed. Defaults to "N". */
	no?: string;
	/** Character that triggers a truthy state when pressed. Defaults to "y". */
	yes?: string;
}

/**
 * A React component that renders a yes/no confirmation prompt.
 */
export function Confirm(props: ConfirmProps) {
	const { invalidError, onSubmit, no = 'N', yes = 'y', ...restProps } = props;
	const [value, setValue] = useState<boolean | null>(null);
	const { isFocused } = useFocus({ autoFocus: true });

	useEffect(() => {
		if (no.length !== 1) {
			console.error(`Prop "no" must be a single character.`);
		} else if (yes.length !== 1) {
			console.error(`Prop "yes" must be a single character.`);
		}
	}, [no, yes]);

	const handleInput = useCallback(
		(input: string) => {
			if (input === no) {
				setValue(false);
				onSubmit(false);
			} else if (input === yes) {
				setValue(true);
				onSubmit(true);
			} else {
				setValue(null);

				throw new Error(invalidError ?? msg('prompt:confirmInvalidValue', { no, yes }));
			}

			// Trigger submit
			return true;
		},
		[invalidError, no, onSubmit, yes],
	);

	return (
		<Prompt<boolean>
			{...restProps}
			afterLabel={
				value === null ? <Style type="muted">{`(${yes}/${no})`}</Style> : <Selected value={value} />
			}
			focused={isFocused}
			value={value}
			onInput={handleInput}
		/>
	);
}
