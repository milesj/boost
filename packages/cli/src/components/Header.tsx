import React from 'react';
import { Box } from 'ink';
import { SPACING_ROW } from '../constants';
import { StyleType } from '../types';
import { Style } from './Style';

export interface HeaderProps {
	/** Text to display for the label. */
	label: string;
	/** The bottom margin. Defaults to `1`. */
	marginTop?: number;
	/** The top margin. Defaults to `1`. */
	marginBottom?: number;
	/** Customize the background color using Style. Defaults to normal text. */
	type?: StyleType | 'none';
}

/**
 * A React component that renders a header/title within a terminal menu.
 */
export function Header(props: HeaderProps) {
	const { label, marginTop = SPACING_ROW, marginBottom = SPACING_ROW, type = 'default' } = props;

	return (
		<Box marginBottom={marginBottom} marginTop={marginTop}>
			<Style bold inverted type={type}>
				{` ${label.toLocaleUpperCase()} `}
			</Style>
		</Box>
	);
}
