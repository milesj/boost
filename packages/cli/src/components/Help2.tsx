import React from 'react';
import { Text } from 'ink';

export function Help() {
  return <Box flexDirection="column"></Box>;
}
