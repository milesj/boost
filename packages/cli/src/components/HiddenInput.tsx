import React from 'react';
import { Input, InputProps } from './Input';

export type HiddenInputProps = Omit<InputProps, 'mask'>;

export function HiddenInput(props: HiddenInputProps) {
  return <Input {...props} mask="" />;
}
