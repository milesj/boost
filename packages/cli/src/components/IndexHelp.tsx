import React from 'react';
import { Box, Text } from 'ink';
import { SPACING_ROW } from '../constants';
import { ProgramOptions } from '../types';
import { Style } from './Style';

export interface IndexHelpProps extends ProgramOptions {
	children?: React.ReactNode;
}

/**
 * A React component that renders a help menu for the entire program.
 */
export function IndexHelp(props: IndexHelpProps) {
	const { banner, bin, children, footer, header, name, version } = props;

	return (
		<Box flexDirection="column" marginBottom={1}>
			{!!banner && (
				<Box>
					<Text>{banner}</Text>
				</Box>
			)}

			<Box marginTop={SPACING_ROW}>
				<Text>
					{`${name} v${version}`} <Style type="muted">{bin}</Style>
				</Text>
			</Box>

			{!!header && (
				<Box marginTop={SPACING_ROW}>
					<Text>{header}</Text>
				</Box>
			)}

			{children}

			{!!footer && (
				<Box marginTop={SPACING_ROW}>
					<Text>{footer}</Text>
				</Box>
			)}
		</Box>
	);
}
