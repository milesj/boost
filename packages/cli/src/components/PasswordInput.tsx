import React from 'react';
import { Input, InputProps } from './Input';

export type PasswordInputProps = Omit<InputProps, 'mask'>;

/**
 * A React component that renders an input field with characters masked with "*".
 */
export function PasswordInput(props: PasswordInputProps) {
	return <Input {...props} mask="*" />;
}
