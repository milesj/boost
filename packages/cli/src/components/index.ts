export * from './Confirm';
export * from './Failure';
export * from './Header';
export * from './Help';
export * from './HiddenInput';
export * from './IndexHelp';
export * from './Input';
export * from './MultiSelect';
export * from './PasswordInput';
export * from './Select';
export * from './Style';
