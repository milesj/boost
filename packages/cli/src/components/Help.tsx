/* eslint-disable react/destructuring-assignment */

import React from 'react';
import { Box, Text } from 'ink';
import { OptionConfig, OptionConfigMap, ParamConfig, ParamConfigList } from '@boost/args';
import { toArray } from '@boost/common';
import { stripAnsi } from '@boost/terminal';
import { DELIMITER, SPACING_COL, SPACING_COL_WIDE, SPACING_ROW } from '../constants';
import {
	formatCommandCall,
	formatDescription,
	formatType,
	getLongestWidth,
	groupByCategory,
} from '../helpers';
import { msg } from '../translate';
import { Categories, CommandConfig, CommandConfigMap } from '../types';
import { Header } from './Header';
import { Style } from './Style';

export interface HelpProps {
	categories?: Categories;
	config?: CommandConfig;
	commands?: CommandConfigMap;
	delimiter?: string;
	header?: string;
	options?: OptionConfigMap;
	params?: ParamConfigList;
}

export class Help extends React.Component<HelpProps> {
	gatherOptionTags(config: OptionConfig): string[] {
		const tags: string[] = [];

		if (config.count) {
			tags.push(msg('cli:tagCount'));
		}

		if (config.multiple) {
			tags.push(msg('cli:tagMultiple') + (config.arity ? `(${config.arity})` : ''));
		}

		return tags;
	}

	gatherParamTags(config: ParamConfig): string[] {
		const tags: string[] = [];

		if (config.required) {
			tags.push(msg('cli:tagRequired'));
		}

		return tags;
	}

	renderCommands(commands: CommandConfigMap) {
		let pathWidth = 0;

		const items = Object.entries(commands).map(([path, command]) => {
			const pathCall = formatCommandCall(path, command);
			const pathCallStripped = stripAnsi(pathCall);

			if (pathCallStripped.length > pathWidth) {
				pathWidth = pathCallStripped.length;
			}

			return {
				...command,
				name: path,
				path,
				pathCall,
			};
		});

		const categorizedCommands = groupByCategory(items, this.props.categories ?? {});
		const categoryCount = Object.keys(categorizedCommands).length;

		return (
			<>
				<Header label={msg('cli:labelCommands')} />

				{Object.entries(categorizedCommands).map(([key, category], index) => (
					<Box key={key} flexDirection="column">
						{!!category.name && categoryCount > 1 && (
							<Box marginTop={index === 0 ? 0 : SPACING_ROW} paddingLeft={SPACING_COL}>
								<Style bold type="none">
									{category.name}
								</Style>
							</Box>
						)}

						{category.items.map((config) => {
							const desc = formatDescription(config);

							return (
								<Box key={key + config.path} flexDirection="row" paddingLeft={SPACING_COL}>
									<Box flexGrow={0} width={pathWidth + SPACING_COL_WIDE}>
										<Text>{config.pathCall}</Text>
									</Box>

									<Box flexGrow={1}>
										<Text wrap="wrap">{desc}</Text>
									</Box>
								</Box>
							);
						})}
					</Box>
				))}
			</>
		);
	}

	renderOptions(options: OptionConfigMap) {
		let longWidth = 0;
		let shortWidth = 0;

		const items = Object.entries(options).map(([name, option]) => {
			const shortLabel = option.short ? `-${option.short},` : '';
			let longLabel = `--${option.default === true ? 'no-' : ''}${name}`;

			if (option.type !== 'boolean') {
				longLabel += ' ';
				longLabel += formatType(option);
			}

			const longLabelStripped = stripAnsi(longLabel);
			const shortLabelStripped = stripAnsi(shortLabel);

			if (longLabelStripped.length > longWidth) {
				longWidth = longLabelStripped.length;
			}

			if (shortLabelStripped.length > shortWidth) {
				shortWidth = shortLabelStripped.length;
			}

			return {
				...option,
				longLabel,
				name,
				shortLabel,
			};
		});

		const categorizedOptions = groupByCategory(items, this.props.categories ?? {});
		const categoryCount = Object.keys(categorizedOptions).length;
		const showShortColumn = shortWidth > 0;

		return (
			<>
				<Header label={msg('cli:labelOptions')} />

				{Object.entries(categorizedOptions).map(([key, category], index) => (
					<Box key={key} flexDirection="column">
						{!!category.name && categoryCount > 1 && (
							<Box marginTop={index === 0 ? 0 : SPACING_ROW} paddingLeft={SPACING_COL}>
								<Style bold type="none">
									{category.name}
								</Style>
							</Box>
						)}

						{category.items.map((config) => (
							<Box key={key + config.name} flexDirection="row" paddingLeft={SPACING_COL}>
								{showShortColumn && (
									<Box flexGrow={0} width={shortWidth + SPACING_COL}>
										<Text>{config.shortLabel}</Text>
									</Box>
								)}

								<Box flexGrow={0} width={longWidth + SPACING_COL_WIDE}>
									<Text>{config.longLabel}</Text>
								</Box>

								<Box flexGrow={1}>
									<Text wrap="wrap">
										{formatDescription(config, this.gatherOptionTags(config))}
									</Text>
								</Box>
							</Box>
						))}
					</Box>
				))}
			</>
		);
	}

	renderParams(params: ParamConfigList) {
		const labels = params.map((config, index) => `${config.label ?? index} ${formatType(config)}`);
		const labelWidth = getLongestWidth(labels);
		const allowVariadic = this.props.config?.allowVariadicParams;

		return (
			<>
				<Header label={msg('cli:labelParams')} />

				{params.map((config, index) => {
					if (config.hidden) {
						return null;
					}

					const desc = config ? formatDescription(config, this.gatherParamTags(config)) : '';

					return (
						<Box key={`${labels[index]}-${index}`} flexDirection="row" paddingLeft={SPACING_COL}>
							<Box flexGrow={0} width={labelWidth + SPACING_COL_WIDE}>
								<Text>{labels[index]}</Text>
							</Box>

							<Box flexGrow={1}>
								<Text wrap="wrap">{desc}</Text>
							</Box>
						</Box>
					);
				})}

				{allowVariadic && (
					<Box paddingLeft={SPACING_COL}>
						<Text>
							{`â€¦${typeof allowVariadic === 'string' ? allowVariadic : ''}`}{' '}
							{formatType({
								multiple: true,
								type: 'string',
							})}
						</Text>
					</Box>
				)}
			</>
		);
	}

	renderUsage(usage: string[] | string = '') {
		const { delimiter = DELIMITER } = this.props;

		return (
			<>
				<Header label={msg('cli:labelUsage')} />

				{toArray(usage).map((example) => (
					<Box key={example} paddingLeft={SPACING_COL}>
						<Text>{example.startsWith(delimiter) ? example : delimiter + example}</Text>
					</Box>
				))}
			</>
		);
	}

	// eslint-disable-next-line complexity
	render() {
		const { commands, options, header, params, config } = this.props;
		const hasDesc = Boolean(config?.description);
		const hasUsage = Boolean(config?.usage && config.usage.length > 0);
		const hasParams = Boolean(params && params.length > 0);
		const hasCommands = Boolean(commands && Object.keys(commands).length > 0);
		const hasOptions = Boolean(options && Object.keys(options).length > 0);

		return (
			<Box flexDirection="column">
				{(hasDesc || hasUsage) && header && <Header label={header} />}

				{hasDesc && (
					<Box paddingLeft={SPACING_COL}>
						<Text>{formatDescription(config!)}</Text>
					</Box>
				)}

				{hasUsage && this.renderUsage(config?.usage)}

				{hasParams && this.renderParams(params!)}

				{hasCommands && this.renderCommands(commands!)}

				{hasOptions && this.renderOptions(options!)}
			</Box>
		);
	}
}
