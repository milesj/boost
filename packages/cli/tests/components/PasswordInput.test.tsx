import React from 'react';
import { render } from 'ink-testing-library';
import { PasswordInput, PasswordInputProps } from '../../src/components/PasswordInput';

describe('PasswordInput', () => {
  const props: PasswordInputProps = {
    label: 'Password?',
    onSubmit() {},
  };

  it('renders label by default', () => {
    const { lastFrame } = render(<PasswordInput {...props} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders placeholder if no default value', () => {
    const { lastFrame } = render(<PasswordInput {...props} placeholder="<password>" />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders default value as stars', () => {
    const { lastFrame } = render(<PasswordInput {...props} defaultValue="test" />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders entered text as stars', async () => {
    const { lastFrame, stdin } = render(<PasswordInput {...props} />);

    await delay();
    stdin.write('hello there');
    await delay();

    expect(lastFrame()).toMatchSnapshot();
  });

  it('calls `onChange` with unmasked value', async () => {
    const spy = jest.fn();
    const { stdin } = render(<PasswordInput {...props} onChange={spy} />);

    await delay();
    stdin.write('test');
    await delay();

    expect(spy).toHaveBeenCalledWith('test');
  });

  it('calls `onSubmit` with unmasked value when pressing return', async () => {
    const spy = jest.fn();
    const { stdin } = render(<PasswordInput {...props} onSubmit={spy} />);

    await delay();
    stdin.write('test');
    await delay();
    stdin.write('\r');
    await delay();

    expect(spy).toHaveBeenCalledWith('test');
  });
});
