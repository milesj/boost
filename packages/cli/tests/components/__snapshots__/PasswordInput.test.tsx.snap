// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput renders default value as stars 1`] = `"[36m?[39m [1mPassword?[22m ****"`;

exports[`PasswordInput renders entered text as stars 1`] = `"[36m?[39m [1mPassword?[22m ***********[46m[30m [39m[49m"`;

exports[`PasswordInput renders label by default 1`] = `"[36m?[39m [1mPassword?[22m"`;

exports[`PasswordInput renders placeholder if no default value 1`] = `"[36m?[39m [1mPassword?[22m [90m<password>[39m"`;
